### PR TITLE
driver: adc: add nxp mcxn10 adc support 

### PR DESCRIPTION
**Pull request by ZhaoxiangJin**
add nxp mcxn10 adc support, the local test has passed

### CODE REVIEW

File: boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml:19]: This should be `adc`, not the name of the device IP

File: dts/arm/nxp/nxp_mcxn94x_common.dtsi
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_mcxn94x_common.dtsi:539]: register style is lowercase, IE `10dd00`. Same thing applies to the `reg` property.

File: samples/drivers/adc/boards/frdm_mcxn947.overlay
Comment (danieldegrasse) [inline samples/drivers/adc/boards/frdm_mcxn947.overlay:28]: Please fix indentation here

File: boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.dts
Comment (danieldegrasse) [inline boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.dts:1]: We should also add the ADC to the list of supported features in the board's `index.rst` file: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/frdm_mcxn947/doc/index.rst?plain=1#L61

**Review by danieldegrasse — COMMENTED** [review]
Thanks for the contribution here- Just a heads up, you'll need to rebase this PR to resolve merge conflicts

File: boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml
Comment (ZhaoxiangJin) [reply boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml:19]: Thanks for your reminder, I will fix it.

File: dts/arm/nxp/nxp_mcxn94x_common.dtsi
Comment (ZhaoxiangJin) [reply dts/arm/nxp/nxp_mcxn94x_common.dtsi:539]: Thanks for your reminder, I will change it to lowercase.

File: boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.dts
Comment (ZhaoxiangJin) [reply boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.dts:1]: Thanks for your reminder, I will add it.

**Comment by ZhaoxiangJin** [comment]
This PR is blocked by https://github.com/zephyrproject-rtos/hal_nxp/pull/358, the MCXN10 LPADC needs vref to provide bias current.

**Comment by danieldegrasse** [comment]
> This PR is blocked by https://github.com/zephyrproject-rtos/hal_nxp/pull/358, the MCXN10 LPADC needs vref to provide bias current.

Take a look at https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules. You should update the west YAML to reference the PR in the HAL. Once that PR merges, you can change the west YAML to reference the new SHA.

For example:

```
 - name: hal_nxp
      revision: pull/358/head
      path: modules/hal/nxp
      groups:
        - hal
 ```

**Comment by ZhaoxiangJin** [comment]
> > This PR is blocked by [zephyrproject-rtos/hal_nxp#358](https://github.com/zephyrproject-rtos/hal_nxp/pull/358), the MCXN10 LPADC needs vref to provide bias current.
> 
> Take a look at https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules. You should update the west YAML to reference the PR in the HAL. Once that PR merges, you can change the west YAML to reference the new SHA.
> 
> For example:
> 
> ```
>  - name: hal_nxp
>       revision: pull/358/head
>       path: modules/hal/nxp
>       groups:
>         - hal
> ```

Yes, I forgot this, updated.

**Review by decsny — CHANGES_REQUESTED** [review]
The VREF already has a driver in zephyr, you should use that through the regulator api

**Comment by ZhaoxiangJin** [comment]
> The VREF already has a driver in zephyr, you should use that through the regulator api

> The VREF already has a driver in zephyr, you should use that through the regulator api

Hi @decsny, thanks for your reply, I have some opinions about this.
1. The LPADC on only a few platforms requires vref to provide a bias current, so enabling vref to provide a bias current should be done in board.c (platform specific).
2. Based on the above point of view, I think it may be troublesome to use the regulator API to set vref in board.c, and it is not as efficient as directly calling the vref driver API in hal-nxp. I think the low-level code (For the vast majority of users, this code is not something they consider) like board.c is different from the application layer code and should pays great attention to efficiency.

**Comment by danieldegrasse** [comment]
> The LPADC on only a few platforms requires vref to provide a bias current, so enabling vref to provide a bias current should be done in board.c (platform specific).

Ideally this isn't something we should do in the board.c file. Instead, we should be using the regulator API within the LPADC driver. The benefit this brings is abstraction- imagine if the user wanted to use the LPADC on a platform where the VREF pins can be connected to an external source, and wanted to drive the VREF pins from a PMIC? If the LPADC driver enabled an attached regulator (defined in devicetree) via the regulator API, this use case would be possible as well as the vref case described here.

Hopefully this example helps explain why abstracting the VREF behind the regulator API is useful

**Comment by decsny** [comment]
There is already code in the zephyr lpadc driver for this purpose that uses the regulator api.

**Comment by ZhaoxiangJin** [comment]
> > The LPADC on only a few platforms requires vref to provide a bias current, so enabling vref to provide a bias current should be done in board.c (platform specific).
> 
> Ideally this isn't something we should do in the board.c file. Instead, we should be using the regulator API within the LPADC driver. The benefit this brings is abstraction- imagine if the user wanted to use the LPADC on a platform where the VREF pins can be connected to an external source, and wanted to drive the VREF pins from a PMIC? If the LPADC driver enabled an attached regulator (defined in devicetree) via the regulator API, this use case would be possible as well as the vref case described here.
> 
> Hopefully this example helps explain why abstracting the VREF behind the regulator API is useful

Hi Daniel, thanks for your comments, your perspective is pretty good. But I feel confused about the **imagine if the user wanted to use the LPADC on a platform where the VREF pins can be connected to an external source, and wanted to drive the VREF pins from a PMIC?**.
According to my understanding, if the user uses the VREFO (VREF output) as the LPADC reference voltage source, the reference voltage from the VREFO is internally connected to the LPADC, so, I don't know what the meaning of the **VREF pins can be connected to an external source, and wanted to drive the VREF pins from a PMIC?**, the VREF on NXP devices (like LPC55s36, KW45, and MCXN10) doesn't support driver from other modules. Maybe I lost some information?
A note point is, in the beginning, I said **provide a bias current**, but that does not mean provide a reference voltage. On some platforms, even if VREFO is not used as the reference voltage, VREF still needs to be turned on to provide a bias current to the LPADC. Take the MCXN10 as an example, you can see in the .dtsi file, the reference voltage of LPADC is configured to 3, which means the reference voltage is VDD_ANA (this information can be found in MCXN10 RM), if you don't open VREF and configure it at 'kVREF_ModeBandgapOnly', the LPADC cannot work correctly.
That is why I think this operation (using the VREF to provide the bias current) needs to be done on the board.c file.
image

**Comment by Albort12138** [comment]
> > The LPADC on only a few platforms requires vref to provide a bias current, so enabling vref to provide a bias current should be done in board.c (platform specific).
> 
> Ideally this isn't something we should do in the board.c file. Instead, we should be using the regulator API within the LPADC driver. The benefit this brings is abstraction- imagine if the user wanted to use the LPADC on a platform where the VREF pins can be connected to an external source, and wanted to drive the VREF pins from a PMIC? If the LPADC driver enabled an attached regulator (defined in devicetree) via the regulator API, this use case would be possible as well as the vref case described here.
> 
> Hopefully this example helps explain why abstracting the VREF behind the regulator API is useful

Based on my understanding, one clarification about ADC reference voltage, it may come from the *VREFH* pin which is an input pin(May from PMIC). It may also come from the *VREF* which is an internal module and is connected to ADC's reference voltage directly. The pin of VREF is an output signal.
For this topic, I agree with Daniel to use a regular API of VREF in ADC driver. Two reasons: 1. If vref is enabled in board layer, it may increase power consumption when the ADC is not used. 2. Vref regular driver support set vref as different mode, it is very flexible. If VREF is used as ADC's reference voltage it can be set as high power mode, if VREF only used to provide bias current it can be set as standby mode

**Comment by ZhaoxiangJin** [comment]
> > > The LPADC on only a few platforms requires vref to provide a bias current, so enabling vref to provide a bias current should be done in board.c (platform specific).
> > 
> > 
> > Ideally this isn't something we should do in the board.c file. Instead, we should be using the regulator API within the LPADC driver. The benefit this brings is abstraction- imagine if the user wanted to use the LPADC on a platform where the VREF pins can be connected to an external source, and wanted to drive the VREF pins from a PMIC? If the LPADC driver enabled an attached regulator (defined in devicetree) via the regulator API, this use case would be possible as well as the vref case described here.
> > Hopefully this example helps explain why abstracting the VREF behind the regulator API is useful
> 
> Based on my understanding, one clarification about ADC reference voltage, it may come from the _VREFH_ pin which is an input pin(May from PMIC). It may also come from the _VREF_ which is an internal module and is connected to ADC's reference voltage directly. The pin of VREF is an output signal. For this topic, I agree with Daniel to use a regular API of VREF in ADC driver. Two reasons: 1. If vref is enabled in board layer, it may increase power consumption when the ADC is not used. 2. Vref regular driver support set vref as different mode, it is very flexible. If VREF is used as ADC's reference voltage it can be set as high power mode, if VREF only used to provide bias current it can be set as standby mode

Good point, some low-power applications may init and deinit LPADC multiple times to achieve low power consumption, so enabling and disabling VREF should also be placed in the LPADC driver. Suppose the VREF is enabled on the board.c file, the VREF will always be open, which will consume more power if LPADC is no longer used.

**Comment by ZhaoxiangJin** [comment]
close this PR, because the adc is already enabled internally on the mcxn947 platform.

**Comment by ZhaoxiangJin** [comment]
Aligned with @mmahadevan108, re-open this PR to enable LPADC for MCXN947 platform in upstream

**Comment by ZhaoxiangJin** [comment]
close this PR, and merge the modified content in this PR into the VREF enablement PR (https://github.com/zephyrproject-rtos/zephyr/pull/70642). Sorry for any review inconvenience.

**Pull request closed without merging**
